File: Dynamic Programming/fibonacci_dp.java
import java.util.*;

public class fibonacci_dp {
    public static void main(String[] args) {
        System.out.println("Enter the number that want to finf=d its fibonacci");
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] dp = new int[n+1];
        Arrays.fill(dp, -1);

        //Tabulization Method
        dp[0] = 0;
        dp[1] = 1;
        for(int i =2; i<=n;i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        System.out.println("The Answer is " + dp[n]);
        

        //Space Optimization
        int prev = 1;
        int prev2 = 0;

        for(int i =2; i<= n; i++){
            int curri = prev + prev2;
            prev2 = prev;
            prev = curri;

        }
        System.out.println("The ANswer is "+ prev);
        
    }
    public static int fibo(int n , int [] dp){
        //Base case
        if(n<=1) return n;
        if(dp[n]!=-1){
            return dp[n];
        }
        return dp[n] = fibo(n-1, dp) + fibo(n-2, dp);
    }

    
}

File: Dynamic Programming/fibonacci_dp.cpp
#include <bits/stdc++.h>
using namespace std;
// Fibonacci function

int f(int n, vector<int> &dp)
{
    // Base casea
    if (n <= 1) return n;
    
    if (dp[n] != -1)
        return dp[n];
    return dp[n] = f(n - 1, dp) + f(n - 2, dp);
}

//<ain funtion
int main()
{
    int n;
    cin >> n;

    vector<int> dp(n + 1, -1);
    cout << f(n, dp);

    //Tabukization
    dp[0] = 0;
     dp[1] = 1;
     for(int i = 2; i<= n ; i++){
        dp[i] = dp[i-1] + dp[i-2];

     }
     printf(" the answer is \n" );
     cout<<dp[n];

     //Space optimazation
     int prev = 1;
     int prev2 = 0;

     for(int i =2; i<= n ; i++){
        int curri = prev + prev2;
        prev2 = prev;
        prev = curri;
     }
     printf("The space optimized result is \n");
     cout<< prev;


    return 0;
}